bitcoin/src/pow.rs

```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+// TODO: this documentation is just a place holder, all the
+// docs need be rewritten when we converge on a design
+/// Error type while get a new block by height
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
+
+impl fmt::Display for GetBlockHeaderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            Self::Generic(error) => error.clone(),
+            Self::NotFound => String::from("Block header not found"),
+        };
+        write!(f, "{msg}")
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for GetBlockHeaderError {}
+
+// FIXME: we should move this in a better place
+/// TODO: add some docs here
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensius_params: Params,
+    last_height: Height,
+    get_block_header_by_heigh: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: Fn(u32) -> Result<Header, GetBlockHeaderError>,
+{
+    let adjustment_interval = consensius_params.difficulty_adjustment_interval();
+    let pow_target_timespan = consensius_params.pow_target_timespan;
+    let no_pow_retargeting = consensius_params.no_pow_retargeting;
+    let last_adjustment_height = last_height.to_consensus_u32() as u64 - (adjustment_interval - 1);
+    let last_adjustment_block = get_block_header_by_heigh(last_adjustment_height as u32)?;
+    let last_adjustment_time = last_adjustment_block.time;
+    let current_height: u64 = last_height.to_consensus_u32() as u64;
+
+    if (current_height % adjustment_interval.min(1)) != 0 {
+        return Ok(last_adjustment_block.target().to_compact_lossy().to_consensus());
+    }
+
+    if no_pow_retargeting {
+        return Ok(last_adjustment_block.bits.to_consensus());
+    }
+
+    let last_time = last_adjustment_block.time;
+    let actual_timespan = (last_time - last_adjustment_time) as u64;
+    let mut adjusted_timespan = actual_timespan;
+
+    if actual_timespan < pow_target_timespan / 4 {
+        adjusted_timespan = pow_target_timespan / 4;
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        return Ok(U256::from(pow_target_timespan).mul_u64(4).0.bits());
+    }
+
+    let target = last_adjustment_block.target();
+    let (mut target, overflow) = target.mul_u64(adjusted_timespan);
+    // FIXME: improve the API of the target.
+    target = Target(target.0 / U256::from(pow_target_timespan));
```

Comment:
Why are you dividing the target by "the number of seconds in 2 weeks"?

Reply:
For this reason https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L66

I am missing something?

Reply:
Copying the bad habits of the original source code is not a requirement. It's much more difficult in C++ to change the name of variables a lot, so they recycle the same variable a lot. This is not needed in Rust.

`target` is no longer a target, but it's a stand-in for a `U256` here that represents the mid-state of a calculation.

You are basically multiplying the target by the ratio of `adjusted_timespan / pow_target_timespan`

This calculation is ok, but the variable names should change between mid-state and final result.

Re-using the `target` variable is confusing.